File: sorting-searching/FerriesWheel.java
import java.util.Arrays;

public class FerriesWheel {
    public static int getMaximumGondolas(int[] weights , int maxWeight){
        int n = weights.length;
        int leftPtr =0;
        int rightPtr = n-1;
        int count = 0;
        Arrays.sort(weights);
        while(leftPtr <= rightPtr){
            if(weights[leftPtr] + weights[rightPtr] <= maxWeight){
                leftPtr++;
                rightPtr--;
            }else{
                rightPtr--;
            }
            count++;
        }
        return count;
    }
    public static void main(String[] args) {
        int maxWeight = 100;
        int[] weights = {90,20,80,30,70,40,60,50,10,100};
        System.out.println(getMaximumGondolas(weights, maxWeight));
    }
}

File: sorting-searching/ferries_wheel.cpp
#include<iostream>
#include<vector>
#include<algorithm>

int get_minimum_gondolas(std::vector<int>&weights , int max_allowed_weight){
    int cnt = 0;
    int n = weights.size();
    std::sort(weights.begin(),weights.end());
    int left_ptr = 0;
    int right_ptr = n-1;
    while(left_ptr <= right_ptr){
        if(weights[left_ptr] + weights[right_ptr] <= max_allowed_weight){
            left_ptr++;
            right_ptr--;
        }else{
            right_ptr--;
        }
        cnt++;
    }
    return cnt;
}

int main(){
    int n,x;
    std::cin>>n>>x;
    std::vector<int>weights(n,0);
    // take input the weights
    for(int i =0;i<n;i++){
        std::cin>>weights[i];
    }
    std::cout<<get_minimum_gondolas(weights,x);
    return 0;
}